### ISSUE TITLE
Different results from the paper (AFW dataset)

### ISSUE THREAD

**Issue by invigen** [natanielruiz/deep-head-pose#9]
Hi, thank you for your great work.
I checked the performance of the pretrained model (300W-LP, alpha 1 and 300W-LP, alpha 2), the result on AFLW2000 are the same as the paper.
But when I test the models on the AFW dataset, the results are very different. 
I wrote my own code to compute the discrete predictions that rounds of the the nearest 15 degree, the yaw accuracy is only 21.15%, and I also check the mean absolute error is 53.3674.
So even if I made some mistake calculating the discrete predictions, the MAE of yaw seems too large. I am wondering which step was missing to reproduce the result in the paper. (I have make sure that the input format are the same as the one required in the datasets.py)

**Comment by natanielruiz** [comment]
Hi, can you try using -yaw instead of yaw for evaluation? The format of the afw annotations are different. Also I’ll upload my testing code for afw soon.

**Comment by invigen** [comment]
Thanks for the quick response.
After I inverted the sign of all yaw labels, the accuracy is still only 67.31% (nearest 15 degree).
I think that I am missing some part of the implementation details.
Will look into the details more carefully while waiting for your testing code.
Thank you in advance!

**Comment by natanielruiz** [comment]
Sorry I still haven't got the chance to upload the testing code but I verified that I obtain the correct paper results. Give me another week and it'll be up!

**Comment by omersumer** [comment]
Hi, I tested three pre-trained models (hopenet_alpha1, hopenet_alpha2, hopenet_robust_alpha1) using "test_hopenet,py", the results that I got are not exactly same as the paper:

Hopenet Alpha-1               Yaw: 7.5806, Pitch: 9.0066, Roll: 8.4781
Hopenet Alpha-2               Yaw: 7.0944, Pitch: 8.8771, Roll: 8.2443
Hopenet Robust Alpha-1   Yaw: 9.7017, Pitch: 9.3551, Roll: 8.5403

And, training script is not working. In paper, learning parameters/solver details are not written. I just tried with default configs in hopenet_train.py, but loss is not minimized. Can you write learning parameters (possible issue in training script) and/or share the original training script that you used, please?

**Comment by natanielruiz** [comment]
The training script is now working. Also, if you look closely you can find the training parameters in the paper. The training script that is shared is the training script used in the paper.

The results that you report are on which dataset?

**Comment by natanielruiz** [comment]
Also sorry about AFW evaluation. I just started a new job and don't have too much time to put out the script but I will try in the coming weeks.

**Comment by omersumer** [comment]
I am sorry for confusion, I missed the solver details in the paper and also wrote under AFW issue.

To clarify, I downloaded three models that you provided and tested using "test_hopenet.py" with following parameters on AFLW2000:
python test_hopenet.py --gpu 0 --data_dir $AFLW2000DIR --filename_list $FILENAME_LIST --snapshot hopenet_*.pkl --dataset AFLW2000

I tried it again, but models current performances seem worse than paper. My PyTorch version is '0.4.0a0+09cff19'. Could you just test on AFLW2000 to check Table I's results
Multi-Loss ResNet50 (α = 2) 6.470 6.559 5.436 6.155
Multi-Loss ResNet50 (α = 1) 6.920 6.637 5.674 6.410

By the way, I will train on 300W-LP and write here test performance on AFLW2000 and BIWI whether they are on the same scale or deviates from paper in a few days.

**Comment by natanielruiz** [comment]
@omersumer Please train again with the most recent training script, I fixed a bug that was making training unstable using the latest PyTorch version. Yes, please report your results if you can.

**Comment by hminle** [comment]
Hi @invigen, where you can download AFLW2000 and Pose_300W-LP?

**Comment by hminle** [comment]
@natanielruiz , I see your default dataset is Pose_300W_LP, where can I download it?

**Comment by developer-mayuan** [comment]
@hminle http://www.cbsr.ia.ac.cn/users/xiangyuzhu/projects/3DDFA/main.htm

**Comment by natanielruiz** [comment]
@invigen By the way I just noticed this about your question. In the paper we **train a model on AFLW** and then **test it on AFW** to compare to all prior work. This is why, when trying to use one of the pre-trained models that I released that are trained on 300W-LP you don't get the accuracy reported on the paper. I will make an effort to release the AFW testing code and the pre-trained model soon.
